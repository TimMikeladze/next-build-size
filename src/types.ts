export type Pages = Record<string, string[]>

export type BuildManifest = {
  ampDevFiles: any[]
  ampFirstPages: any[]
  devFiles: any[]
  lowPriorityFiles: string[]
  pages: Pages
  polyfillFiles: string[]
  rootMainFiles: any[]
}

export type GetPageSizesArgs = {
  nextDir?: string
}

export type WritePageSizeArgs = {
  nextDir: string
  output: string
}

export type PageSizes = Record<string, string>

export type CachePageSizesArgs = {
  filePath: string
  pageSizes: PageSizes
}
