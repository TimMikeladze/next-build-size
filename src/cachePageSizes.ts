import { CachePageSizesArgs } from './types'
import cache from '@actions/cache'
import { VERSION } from './constants'

const key = 'next-build-size-' + VERSION

export const cachePageSizes = async (args: CachePageSizesArgs) => {
  const paths = [args.filePath]
  const cacheId = await cache.saveCache(paths, key)
}
