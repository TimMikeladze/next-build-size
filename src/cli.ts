#!/usr/bin/env node
import { Command } from 'commander'
import {
  cacheNextBuildSize,
  comparePageSizes,
  getOutputDirectory,
  PageSizes,
  restoreNextBuildSize,
  VERSION,
  writePageSizes
} from './index'
import { writeFileSync } from 'fs'
import { resolve } from 'path'

const sendToDiscord = async (url, body: { content: string }): Promise<void> => {
  await fetch(url, {
    body: JSON.stringify(body),
    headers: {
      'Content-type': 'application/json'
    },
    method: 'POST'
  })
}

const sendToSlack = async (url, body: { text: string }): Promise<void> => {
  await fetch(url, {
    body: JSON.stringify(body),
    headers: {
      'Content-type': 'application/json'
    },
    method: 'POST'
  })
}

export default sendToDiscord
;(async () => {
  const program = new Command()

  program
    .name('next-build-size')
    .description('Get and compare page sizes of a Next.js app')
    .version(VERSION)

  program
    .command('stats')
    .description('Get and compare page sizes')
    .argument(
      '[nextDir]',
      'Optional path to a .next folder, this is typically the root of your Next.js app after running `next build`',
      '.next'
    )
    .option(
      '-o, --output <output>',
      'Path to a folder where the stats will be written to',
      'next-build-size'
    )
    .option('-d, --discord <discord>', 'Send the stats to a Discord webhook')
    .option('-s, --slack <slack>', 'Send the stats to a Slack webhook')
    .option('-t, --title <title>', 'Title of the Discord or Slack message')
    .option(
      '-v, --verbose',
      'Show information about the execution of the command'
    )
    .option(
      '-c, --cache',
      'Save the stats in the Github Actions cache and compare them with the previous build'
    )
    .action(async (nextDir = '.next', options) => {
      const directoryPath = getOutputDirectory(options.output)
      const verbose = !!options.verbose

      let restoredPageSizes: PageSizes = []

      if (options.cache) {
        restoredPageSizes = await restoreNextBuildSize({ directoryPath })
        if (verbose) {
          if (Object.keys(restoredPageSizes).length) {
            console.log('Restored page stats from cache')
          } else {
            console.log('No page stats found in cache')
          }
        }
      }

      const { filePath, pageSizes } = writePageSizes({
        nextDir,
        output: options.output
      })

      if (verbose) {
        console.log(`Wrote page sizes to ${filePath}`)
      }

      if (options.cache) {
        await cacheNextBuildSize({ directoryPath })
        if (verbose) {
          console.log('Saved stats to cache')
        }
      }

      const comparedPageSizes = comparePageSizes(pageSizes, restoredPageSizes)

      const title = ['Next.js Build Stats', options.title]
        .filter(Boolean)
        .join(' - ')

      const statsFilePath = resolve(directoryPath, 'stats.json')

      const stats = JSON.stringify(comparedPageSizes, null, 2)

      writeFileSync(statsFilePath, stats)

      if (verbose) {
        console.log(`Wrote stats to ${statsFilePath}`)
      }

      const message = `**${title}**
\`\`\`json
${stats}
\`\`\``

      if (options.discord) {
        await sendToDiscord(options.discord, { content: message })
        if (verbose) {
          console.log('Sent stats to Discord')
        }
      }

      if (options.slack) {
        await sendToSlack(options.slack, { text: message })
        if (verbose) {
          console.log('Sent stats to Slack')
        }
      }

      console.log(stats)
    })

  await program.parseAsync(process.argv)

  process.exit(0)
})()
