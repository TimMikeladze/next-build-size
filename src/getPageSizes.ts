import fs from 'fs'
import path, { resolve } from 'path'
import zlib from 'zlib'
import { BuildManifest, GetPageSizesArgs, PageSizes } from './types'

export const getPageSizes = (args: GetPageSizesArgs): PageSizes => {
  const nextDirPath = resolve(
    process.cwd(),
    args.nextDir ? args.nextDir : '.next'
  )

  const buildManifest = JSON.parse(
    fs.readFileSync(path.join(nextDirPath, 'build-manifest.json'), 'utf8')
  ) as BuildManifest

  const pageSizes = Object.keys(buildManifest.pages).map((filePath) => {
    const files = buildManifest.pages[filePath]
    const size = files
      .map((filename) => {
        const gzipped = zlib.gzipSync(
          fs.readFileSync(path.join(nextDirPath, filename))
        )
        return gzipped.byteLength
      })
      .reduce((s, b) => s + b, 0)

    return { filePath, size }
  })

  return pageSizes.reduce((acc, { filePath, size }) => {
    acc[filePath] = formatBytes(size)
    return acc
  }, {} as Record<string, string>)
}

const formatBytes = (bytes: number, decimals = 2) => {
  if (!+bytes) return '0 Bytes'

  const k = 1024
  const dm = decimals < 0 ? 0 : decimals
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']

  const i = Math.floor(Math.log(bytes) / Math.log(k))

  return `${parseFloat((bytes / k ** i).toFixed(dm))} ${sizes[i]}`
}
