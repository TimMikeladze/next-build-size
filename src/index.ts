import cache from '@actions/cache'

import fs, {
  existsSync,
  mkdirSync,
  readFileSync,
  unlinkSync,
  writeFileSync
} from 'fs'
import { resolve } from 'path'
import zlib from 'zlib'

export const VERSION = '1.0.9'

const key = 'next-build-size-' + VERSION

export type Pages = Record<string, string[]>

export type BuildManifest = {
  ampDevFiles: any[]
  ampFirstPages: any[]
  devFiles: any[]
  lowPriorityFiles: string[]
  pages: Pages
  polyfillFiles: string[]
  rootMainFiles: any[]
}

export type GetPageSizesArgs = {
  nextDir?: string
}

export type WritePageSizeArgs = {
  nextDir: string
  output: string
}

export type PageSizes = {
  filePath: string
  size: number
}[]

export type CacheNextBuildSizeArgs = {
  directoryPath: string
}

export type RestoreNextBuildSizeArgs = {
  directoryPath: string
}

export type ComparedPageSizes = {
  [key: string]: {
    diff: string
    size: string
  }
}

export const cacheNextBuildSize = async (args: CacheNextBuildSizeArgs) => {
  const paths = [args.directoryPath]
  return await cache.saveCache(paths, key)
}

export const restoreNextBuildSize = async (
  args: RestoreNextBuildSizeArgs
): Promise<PageSizes> => {
  const paths = [args.directoryPath]
  const cacheId = await cache.restoreCache(paths, key)

  if (!cacheId) {
    return []
  }

  const filePath = resolve(args.directoryPath, 'next-build-size.json')

  const data = JSON.parse(readFileSync(filePath, 'utf8'))

  unlinkSync(filePath)

  return data
}

export const getPageSizes = (args: GetPageSizesArgs) => {
  const nextDirPath = resolve(
    process.cwd(),
    args.nextDir ? args.nextDir : '.next'
  )

  const buildManifest = JSON.parse(
    fs.readFileSync(resolve(nextDirPath, 'build-manifest.json'), 'utf8')
  ) as BuildManifest

  let pageSizes = Object.keys(buildManifest.pages).map((filePath) => {
    const files = buildManifest.pages[filePath]
    const size = files
      .map((filename) => {
        const gzipped = zlib.gzipSync(
          fs.readFileSync(resolve(nextDirPath, filename))
        )
        return gzipped.byteLength
      })
      .reduce((s, b) => s + b, 0)

    return { filePath, size }
  })

  pageSizes = pageSizes.sort((a, b) => b.size - a.size)

  return pageSizes
}

export const getOutputDirectory = (path: string) => {
  const directoryPath = resolve(process.cwd(), path)

  if (!existsSync(directoryPath)) {
    mkdirSync(directoryPath)
  }

  return directoryPath
}

export const writePageSizes = (args: WritePageSizeArgs) => {
  const pageSizes = getPageSizes({
    nextDir: args.nextDir
  })

  const directoryPath = getOutputDirectory(args.output)

  const filePath = resolve(directoryPath, 'sizes.json')

  writeFileSync(filePath, JSON.stringify(pageSizes, null, 2))

  return {
    directoryPath,
    filePath,
    pageSizes
  }
}

export const comparePageSizes = (
  currentPages: PageSizes,
  oldPages: PageSizes
): ComparedPageSizes => {
  const compared: ComparedPageSizes = {}

  currentPages.forEach((currentPage) => {
    const oldPage = oldPages.find((b) => b.filePath === currentPage.filePath)

    const diff = oldPage !== undefined ? currentPage.size - oldPage.size : 0

    let sign

    if (diff > 0) {
      sign = '+'
    } else if (diff < 0) {
      sign = '-'
    } else {
      sign = ''
    }

    compared[currentPage.filePath] = {
      diff: `${sign}${formatBytes(Math.abs(diff))}`,
      size: formatBytes(currentPage.size)
    }
  })

  return compared
}

export const formatBytes = (bytes: number, decimals = 2) => {
  if (!+bytes) return '0 Bytes'

  const k = 1024
  const dm = decimals < 0 ? 0 : decimals
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB']

  const i = Math.floor(Math.log(bytes) / Math.log(k))

  return `${parseFloat((bytes / k ** i).toFixed(dm))} ${sizes[i]}`
}
