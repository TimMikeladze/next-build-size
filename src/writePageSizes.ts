import { resolve } from 'path'
import { lstatSync, writeFileSync } from 'fs'
import { getPageSizes } from './getPageSizes'
import { WritePageSizeArgs } from './types'

export const writePageSizes = (args: WritePageSizeArgs) => {
  const pageSizes = getPageSizes({
    nextDir: args.nextDir
  })

  const filePath = resolve(process.cwd(), args.output)

  try {
    const stats = lstatSync(filePath)

    if (stats.isDirectory()) {
      throw new Error(
        `The path ${args.output} is a directory, please provide a file path`
      )
    }
  } catch (error: any) {
    if (error.code !== 'ENOENT') {
      throw error
    }
  }

  writeFileSync(filePath, JSON.stringify(pageSizes, null, 2))

  return {
    filePath,
    pageSizes
  }
}
