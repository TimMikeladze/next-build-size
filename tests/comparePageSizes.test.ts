import { comparePageSizes } from '../src'

describe('comparePageSizes', () => {
  it('works', () => {
    expect(
      comparePageSizes(
        [
          {
            filePath: '_app',
            size: 10000
          },
          {
            filePath: '_error',
            size: 20000
          },
          {
            filePath: '_404',
            size: 20000
          }
        ],
        [
          {
            filePath: '_app',
            size: 10000
          },
          {
            filePath: '_error',
            size: 10000
          },
          {
            filePath: '_404',
            size: 30000
          }
        ]
      )
    ).toMatchSnapshot()
  })
})
