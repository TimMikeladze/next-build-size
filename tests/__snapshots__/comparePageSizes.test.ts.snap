// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comparePageSizes works 1`] = `
{
  "_404": {
    "diff": "-9.77 KB",
    "size": "19.53 KB",
  },
  "_app": {
    "diff": "0 Bytes",
    "size": "9.77 KB",
  },
  "_error": {
    "diff": "+9.77 KB",
    "size": "19.53 KB",
  },
}
`;
